File: go/test/channel_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/funisgo-streaming-sdk/go"
	"github.com/voxgig-sdk/funisgo-streaming-sdk/go/core"

	vs "github.com/voxgig-sdk/funisgo-streaming-sdk/go/utility/struct"
)

func TestChannelEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.Channel(nil)
		if ent == nil {
			t.Fatal("expected non-nil ChannelEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"channel": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.Channel(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.MakeConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.Channel(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := channelBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "channel." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set FUNISGOSTREAMING_TEST_CHANNEL_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		channelRef01Ent := client.Channel(nil)
		channelRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath([]any{"new", "channel"}, setup.data), "channel_ref01"))

		channelRef01DataResult, err := channelRef01Ent.Create(channelRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		channelRef01Data = core.ToMapAny(channelRef01DataResult)
		if channelRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if channelRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		channelRef01Match := map[string]any{}

		channelRef01ListResult, err := channelRef01Ent.List(channelRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		channelRef01List, channelRef01ListOk := channelRef01ListResult.([]any)
		if !channelRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", channelRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(channelRef01List), map[string]any{"id": channelRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		channelRef01DataUp0Up := map[string]any{
			"id": channelRef01Data["id"],
		}

		channelRef01MarkdefUp0Name := "category"
		channelRef01MarkdefUp0Value := fmt.Sprintf("Mark01-channel_ref01_%d", setup.now)
		channelRef01DataUp0Up[channelRef01MarkdefUp0Name] = channelRef01MarkdefUp0Value

		channelRef01ResdataUp0Result, err := channelRef01Ent.Update(channelRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		channelRef01ResdataUp0 := core.ToMapAny(channelRef01ResdataUp0Result)
		if channelRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if channelRef01ResdataUp0["id"] != channelRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if channelRef01ResdataUp0[channelRef01MarkdefUp0Name] != channelRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", channelRef01MarkdefUp0Name, channelRef01ResdataUp0[channelRef01MarkdefUp0Name])
		}

		// LOAD
		channelRef01MatchDt0 := map[string]any{
			"id": channelRef01Data["id"],
		}
		channelRef01DataDt0Loaded, err := channelRef01Ent.Load(channelRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		channelRef01DataDt0LoadResult := core.ToMapAny(channelRef01DataDt0Loaded)
		if channelRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if channelRef01DataDt0LoadResult["id"] != channelRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		channelRef01MatchRm0 := map[string]any{
			"id": channelRef01Data["id"],
		}
		_, err = channelRef01Ent.Remove(channelRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		channelRef01MatchRt0 := map[string]any{}

		channelRef01ListRt0Result, err := channelRef01Ent.List(channelRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		channelRef01ListRt0, channelRef01ListRt0Ok := channelRef01ListRt0Result.([]any)
		if !channelRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", channelRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(channelRef01ListRt0), map[string]any{"id": channelRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func channelBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "channel", "ChannelTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read channel test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse channel test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"channel01", "channel02", "channel03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("FUNISGOSTREAMING_TEST_CHANNEL_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"FUNISGOSTREAMING_TEST_CHANNEL_ENTID": idmap,
		"FUNISGOSTREAMING_TEST_LIVE":      "FALSE",
		"FUNISGOSTREAMING_TEST_EXPLAIN":   "FALSE",
		"FUNISGOSTREAMING_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["FUNISGOSTREAMING_TEST_CHANNEL_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["FUNISGOSTREAMING_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["FUNISGOSTREAMING_APIKEY"],
			},
			extra,
		})
		client = sdk.NewFunisgoStreamingSDK(core.ToMapAny(mergedOpts))
	}

	live := env["FUNISGOSTREAMING_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["FUNISGOSTREAMING_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: rb/test/channel_entity_test.rb
# Channel entity test

require "minitest/autorun"
require "json"
require_relative "../FunisgoStreaming_sdk"
require_relative "runner"

class ChannelEntityTest < Minitest::Test
  def test_create_instance
    testsdk = FunisgoStreamingSDK.test(nil, nil)
    ent = testsdk.Channel(nil)
    assert !ent.nil?
  end

  # Feature #4: the entity stream(action, ...) method runs the op pipeline and
  # returns an Enumerator over result items. With the streaming feature active
  # it yields the feature's incremental output; otherwise it falls back to the
  # materialised list so stream always yields.
  def test_stream
    seed = {
      "entity" => {
        "channel" => {
          "s1" => { "id" => "s1" },
          "s2" => { "id" => "s2" },
          "s3" => { "id" => "s3" },
        },
      },
    }

    # Fallback: streaming inactive -> yields the materialised list items.
    base = FunisgoStreamingSDK.test(seed, nil)
    seen = base.Channel(nil).stream("list", nil, nil).to_a
    assert_equal 3, seen.length

    # Inbound: streaming active -> yields each item from the feature.
    cfg = FunisgoStreamingConfig.make_config
    if cfg["feature"].is_a?(Hash) && cfg["feature"].key?("streaming")
      sdk = FunisgoStreamingSDK.test(seed, { "feature" => { "streaming" => { "active" => true } } })
      got = []
      sdk.Channel(nil).stream("list", nil, nil).each do |item|
        if item.is_a?(Array)
          got.concat(item)
        else
          got << item
        end
      end
      assert_equal 3, got.length
    end
  end

  def test_basic_flow
    setup = channel_basic_setup(nil)
    # Per-op sdk-test-control.json skip.
    _live = setup[:live] || false
    ["create", "list", "update", "load", "remove"].each do |_op|
      _should_skip, _reason = Runner.is_control_skipped("entityOp", "channel." + _op, _live ? "live" : "unit")
      if _should_skip
        skip(_reason || "skipped via sdk-test-control.json")
        return
      end
    end
    # The basic flow consumes synthetic IDs from the fixture. In live mode
    # without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup[:synthetic_only]
      skip "live entity test uses synthetic IDs from fixture — set FUNISGOSTREAMING_TEST_CHANNEL_ENTID JSON to run live"
      return
    end
    client = setup[:client]

    # CREATE
    channel_ref01_ent = client.Channel(nil)
    channel_ref01_data = Helpers.to_map(Vs.getprop(
      Vs.getpath(setup[:data], "new.channel"), "channel_ref01"))

    channel_ref01_data_result = channel_ref01_ent.create(channel_ref01_data, nil)
    channel_ref01_data = Helpers.to_map(channel_ref01_data_result)
    assert !channel_ref01_data.nil?
    assert !channel_ref01_data["id"].nil?

    # LIST
    channel_ref01_match = {}

    channel_ref01_list_result = channel_ref01_ent.list(channel_ref01_match, nil)
    assert channel_ref01_list_result.is_a?(Array)

    found_item = Vs.select(
      Runner.entity_list_to_data(channel_ref01_list_result),
      { "id" => channel_ref01_data["id"] })
    assert !Vs.isempty(found_item)

    # UPDATE
    channel_ref01_data_up0_up = {
      "id" => channel_ref01_data["id"],
    }

    channel_ref01_markdef_up0_name = "category"
    channel_ref01_markdef_up0_value = "Mark01-channel_ref01_#{setup[:now]}"
    channel_ref01_data_up0_up[channel_ref01_markdef_up0_name] = channel_ref01_markdef_up0_value

    channel_ref01_resdata_up0_result = channel_ref01_ent.update(channel_ref01_data_up0_up, nil)
    channel_ref01_resdata_up0 = Helpers.to_map(channel_ref01_resdata_up0_result)
    assert !channel_ref01_resdata_up0.nil?
    assert_equal channel_ref01_resdata_up0["id"], channel_ref01_data_up0_up["id"]
    assert_equal channel_ref01_resdata_up0[channel_ref01_markdef_up0_name], channel_ref01_markdef_up0_value

    # LOAD
    channel_ref01_match_dt0 = {
      "id" => channel_ref01_data["id"],
    }
    channel_ref01_data_dt0_loaded = channel_ref01_ent.load(channel_ref01_match_dt0, nil)
    channel_ref01_data_dt0_load_result = Helpers.to_map(channel_ref01_data_dt0_loaded)
    assert !channel_ref01_data_dt0_load_result.nil?
    assert_equal channel_ref01_data_dt0_load_result["id"], channel_ref01_data["id"]

    # REMOVE
    channel_ref01_match_rm0 = {
      "id" => channel_ref01_data["id"],
    }
    channel_ref01_ent.remove(channel_ref01_match_rm0, nil)

    # LIST
    channel_ref01_match_rt0 = {}

    channel_ref01_list_rt0_result = channel_ref01_ent.list(channel_ref01_match_rt0, nil)
    assert channel_ref01_list_rt0_result.is_a?(Array)

    not_found_item = Vs.select(
      Runner.entity_list_to_data(channel_ref01_list_rt0_result),
      { "id" => channel_ref01_data["id"] })
    assert Vs.isempty(not_found_item)

  end
end

def channel_basic_setup(extra)
  Runner.load_env_local

  entity_data_file = File.join(__dir__, "..", "..", ".sdk", "test", "entity", "channel", "ChannelTestData.json")
  entity_data_source = File.read(entity_data_file)
  entity_data = JSON.parse(entity_data_source)

  options = {}
  options["entity"] = entity_data["existing"]

  client = FunisgoStreamingSDK.test(options, extra)

  # Generate idmap via transform.
  idmap = Vs.transform(
    ["channel01", "channel02", "channel03"],
    {
      "`$PACK`" => ["", {
        "`$KEY`" => "`$COPY`",
        "`$VAL`" => ["`$FORMAT`", "upper", "`$COPY`"],
      }],
    }
  )

  # Detect ENTID env override before envOverride consumes it. When live
  # mode is on without a real override, the basic test runs against synthetic
  # IDs from the fixture and 4xx's. Surface this so the test can skip.
  entid_env_raw = ENV["FUNISGOSTREAMING_TEST_CHANNEL_ENTID"]
  idmap_overridden = !entid_env_raw.nil? && entid_env_raw.strip.start_with?("{")

  env = Runner.env_override({
    "FUNISGOSTREAMING_TEST_CHANNEL_ENTID" => idmap,
    "FUNISGOSTREAMING_TEST_LIVE" => "FALSE",
    "FUNISGOSTREAMING_TEST_EXPLAIN" => "FALSE",
    "FUNISGOSTREAMING_APIKEY" => "NONE",
  })

  idmap_resolved = Helpers.to_map(
    env["FUNISGOSTREAMING_TEST_CHANNEL_ENTID"])
  if idmap_resolved.nil?
    idmap_resolved = Helpers.to_map(idmap)
  end

  if env["FUNISGOSTREAMING_TEST_LIVE"] == "TRUE"
    merged_opts = Vs.merge([
      {
        "apikey" => env["FUNISGOSTREAMING_APIKEY"],
      },
      extra || {},
    ])
    client = FunisgoStreamingSDK.new(Helpers.to_map(merged_opts))
  end

  live = env["FUNISGOSTREAMING_TEST_LIVE"] == "TRUE"
  {
    client: client,
    data: entity_data,
    idmap: idmap_resolved,
    env: env,
    explain: env["FUNISGOSTREAMING_TEST_EXPLAIN"] == "TRUE",
    live: live,
    synthetic_only: live && !idmap_overridden,
    now: (Time.now.to_f * 1000).to_i,
  }
end
